File: apps/client/src/lib/chat/contracts/reads.ts
import type {
  ChatActivityId,
  ChatThreadId,
  ChatThreadState,
  ChatTurnId,
  ChatTurnState,
  ProjectWorkingFolderId,
  ContinuationGroupId,
  ModelId,
  ProviderFamilyId,
  ProviderInstanceId,
  ProviderItemId,
  ProviderSessionId,
  ProviderSessionState,
  ProviderThreadId,
  ProviderTurnId,
  TurnModeSnapshot,
  UtcTimestamp,
  VersionedJson,
} from "./common";
import type { ChangedFileSummary, ThreadUsageUpdatedEvent } from "./events";
import type { ModelOptionSelection, ProviderCapabilities } from "./provider";

export interface ProviderSessionSnapshot {
  sessionId: ProviderSessionId;
  state: ProviderSessionState;
  providerThreadId: ProviderThreadId | null;
  continuationGroupId: ContinuationGroupId;
  resumeCursor: VersionedJson | null;
  effectiveModes: TurnModeSnapshot;
  capabilities: ProviderCapabilities;
  startedAt: UtcTimestamp;
}

export interface TurnDispatchReceipt {
  turnId: ChatTurnId;
  state: ChatTurnState;
  providerTurnId: ProviderTurnId | null;
  acceptedAt: UtcTimestamp;
}

export interface DriverOperationReceipt {
  accepted: boolean;
  operationId: string;
  detail: string | null;
}

export interface ProviderHistoryItem {
  providerItemId: ProviderItemId | null;
  providerTurnId: ProviderTurnId | null;
  kind: string;
  data: VersionedJson;
}

export interface ProviderHistoryPage {
  items: ProviderHistoryItem[];
  nextCursor: string | null;
}

export interface ChatThreadShellRead {
  id: ChatThreadId;
  workingFolderId: ProjectWorkingFolderId;
  projectId: string;
  title: string;
  providerFamilyId: ProviderFamilyId;
  providerInstanceId: ProviderInstanceId;
  providerThreadId: ProviderThreadId | null;
  modelId: ModelId | null;
  modelOptions: ModelOptionSelection[];
  modes: TurnModeSnapshot;
  state: ChatThreadState;
  latestTurnState: ChatTurnState | null;
  latestPreview: string | null;
  messageCount: number;
  revision: number;
  lastEventSequence: number;
  lastActivityAt: UtcTimestamp;
  unreadAt: UtcTimestamp | null;
  archivedAt: UtcTimestamp | null;
}

export interface ChatProjectShellRead {
  projectId: string;
  workingFolderId: ProjectWorkingFolderId;
  workingFolderName: string;
  workingFolderArchivedAt: UtcTimestamp | null;
  activeThreadCount: number;
  archivedThreadCount: number;
}

export interface ChatChangeNotification {
  threadId: ChatThreadId;
  sequence: number;
  revision: number;
  changedProjectionKeys: string[];
}

export interface ChatTimelineItemRead {
  activityId: ChatActivityId;
  turnId: ChatTurnId | null;
  sequenceAnchor: number;
  kind: string;
  data: VersionedJson;
}

export interface ChatTimelineTurnRead {
  turnId: ChatTurnId;
  state: ChatTurnState;
  startedAt: UtcTimestamp | null;
  completedAt: UtcTimestamp | null;
  stopReason: string | null;
  modelId: ModelId | null;
  modelOptions: ModelOptionSelection[];
  modes: TurnModeSnapshot;
  usage: ThreadUsageUpdatedEvent | null;
  changedFiles: ChangedFileSummary[];
}

export interface ChatTimelinePageRead {
  threadId: ChatThreadId;
  items: ChatTimelineItemRead[];
  turns: ChatTimelineTurnRead[];
  previousCursor: string | null;
  nextCursor: string | null;
  threadRevision: number;
}

File: apps/client/src-tauri/src/chat/models/reads.rs
use super::{
    ChatActivityId, ChatThreadId, ChatThreadState, ChatTurnId, ChatTurnState, ContinuationGroupId,
    ModelId, ModelOptionSelection, ProjectWorkingFolderId, ProviderCapabilities, ProviderFamilyId,
    ProviderInstanceId, ProviderSessionId, ProviderSessionState, ProviderThreadId,
    TurnModeSnapshot, UtcTimestamp, VersionedJson,
};
use crate::chat::events::{ChangedFileSummary, ThreadUsageUpdatedEvent};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderSessionSnapshot {
    pub session_id: ProviderSessionId,
    pub state: ProviderSessionState,
    pub provider_thread_id: Option<ProviderThreadId>,
    pub continuation_group_id: ContinuationGroupId,
    pub resume_cursor: Option<VersionedJson>,
    pub effective_modes: TurnModeSnapshot,
    pub capabilities: ProviderCapabilities,
    pub started_at: UtcTimestamp,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TurnDispatchReceipt {
    pub turn_id: ChatTurnId,
    pub state: ChatTurnState,
    pub provider_turn_id: Option<super::ProviderTurnId>,
    pub accepted_at: UtcTimestamp,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DriverOperationReceipt {
    pub accepted: bool,
    pub operation_id: String,
    pub detail: Option<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderHistoryItem {
    pub provider_item_id: Option<super::ProviderItemId>,
    pub provider_turn_id: Option<super::ProviderTurnId>,
    pub kind: String,
    pub data: VersionedJson,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderHistoryPage {
    pub items: Vec<ProviderHistoryItem>,
    pub next_cursor: Option<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatThreadShellRead {
    pub id: ChatThreadId,
    pub working_folder_id: ProjectWorkingFolderId,
    pub project_id: String,
    pub title: String,
    pub provider_family_id: ProviderFamilyId,
    pub provider_instance_id: ProviderInstanceId,
    pub provider_thread_id: Option<ProviderThreadId>,
    pub model_id: Option<ModelId>,
    pub model_options: Vec<ModelOptionSelection>,
    pub modes: TurnModeSnapshot,
    pub state: ChatThreadState,
    pub latest_turn_state: Option<ChatTurnState>,
    pub latest_preview: Option<String>,
    pub message_count: u64,
    pub revision: u64,
    pub last_event_sequence: u64,
    pub last_activity_at: UtcTimestamp,
    pub unread_at: Option<UtcTimestamp>,
    pub archived_at: Option<UtcTimestamp>,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatChangeNotification {
    pub thread_id: ChatThreadId,
    pub sequence: u64,
    pub revision: u64,
    pub changed_projection_keys: Vec<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatTimelineItemRead {
    pub activity_id: ChatActivityId,
    pub turn_id: Option<ChatTurnId>,
    pub sequence_anchor: u64,
    pub kind: String,
    pub data: VersionedJson,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatTimelineTurnRead {
    pub turn_id: ChatTurnId,
    pub state: ChatTurnState,
    pub started_at: Option<UtcTimestamp>,
    pub completed_at: Option<UtcTimestamp>,
    pub stop_reason: Option<String>,
    pub model_id: Option<ModelId>,
    pub model_options: Vec<ModelOptionSelection>,
    pub modes: TurnModeSnapshot,
    pub usage: Option<ThreadUsageUpdatedEvent>,
    pub changed_files: Vec<ChangedFileSummary>,
}

#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatTimelinePageRead {
    pub thread_id: ChatThreadId,
    pub items: Vec<ChatTimelineItemRead>,
    pub turns: Vec<ChatTimelineTurnRead>,
    pub previous_cursor: Option<String>,
    pub next_cursor: Option<String>,
    pub thread_revision: u64,
}
